src/platform/storage/zephyr/nvs/storage.c

```diff
@@ -163,41 +211,85 @@ crc_check(const unsigned char *data, const size_t data_len) {
 
 mender_err_t
 mender_storage_init(void) {
-    int      result;
-    uint32_t n_sectors;
+    int                      result;
+    int                      part_id;
+    uint16_t                 sector_offset;
+    uint32_t                 n_sectors;
+    const struct flash_area *fap;
+    off_t                    part_offset;
+
+#ifndef CONFIG_MENDER_STORAGE_PARTITION_CUSTOM
+    /* With non-custom storage, we already know the number of sectors we need to
+       work with. */
+    n_sectors = sizeof(flash_sectors) / sizeof(flash_sectors[0]);
+
+    if (MENDER_OK != mender_get_storage_spec(&(mender_storage_nvs_handle.flash_device), &part_id, &sector_offset)) {
+        /* This should never happen with a non-custom partition spec and
+           tests/debug builds should fail hard. */
+        assert(false);
+
+        mender_log_error("Failed to get storage spec");
+        return MENDER_FAIL;
+    }
+#else
+    /* With custom storage, we need to call mender_get_storage_spec() to get the
+       sector offset and only then we'll know how many offsets we need to work
+       with. */
+    if (MENDER_OK != mender_get_storage_spec(&(mender_storage_nvs_handle.flash_device), &part_id, &sector_offset)) {
+        mender_log_error("Failed to get custom storage spec");
+        return MENDER_FAIL;
+    }
```

Comment:
Except for the `assert`, this block is the same as above. Consider moving it out of the `#ifndef` / `#else` and adjust a bit the `n_sectors = ...` line